packages/rpc/src/vatidation/BundlerCollectorTracer.ts

```diff
@@ -0,0 +1,365 @@
+                        topLevelTargetAddress,
+                        access: {},
+                        opcodes: {},
+                        extCodeAccessInfo: {},
+                        contractSize: {}
+                    }
+                    this.topLevelCallCounter++
+                } else if (opcode === "LOG1") {
+                    // ignore log data ofs, len
+                    const topic = log.stack.peek(2).toString(16)
+                    if (topic === this.stopCollectingTopic) {
+                        this.stopCollecting = true
+                    }
+                }
+                this.lastOp = ""
+                return
+            }
+
+            const lastOpInfo = this.lastThreeOpcodes[this.lastThreeOpcodes.length - 2]
+            // store all addresses touched by EXTCODE* opcodes
+            if (lastOpInfo && lastOpInfo.opcode && lastOpInfo.opcode.match(/^(EXT.*)$/) != null) {
+                const addr = toAddress(lastOpInfo.stackTop3[0].toString(16))
+                const addrHex = toHex(addr)
+                const last3opcodesString = this.lastThreeOpcodes.map((x) => x.opcode).join(" ")
+                // only store the last EXTCODE* opcode per address - could even be a boolean for our current use-case
+                // [OP-051]
+                if (last3opcodesString.match(/^(\w+) EXTCODESIZE ISZERO$/) == null) {
+                    this.currentLevel.extCodeAccessInfo[addrHex] = opcode
+                    // this.debug.push(`potentially illegal EXTCODESIZE without ISZERO for ${addrHex}`)
+                } else {
+                    // this.debug.push(`safe EXTCODESIZE with ISZERO for ${addrHex}`)
+                }
+            }
+
+            // not using 'isPrecompiled' to only allow the ones defined by the ERC-4337 as stateless precompiles
+            // [OP-062]
+            const isAllowedPrecompiled: (address: any) => boolean = (address) => {
+                const addrHex = toHex(address)
+                const addressInt = parseInt(addrHex)
+                // this.debug.push(`isPrecompiled address=${addrHex} addressInt=${addressInt}`)
+                return addressInt > 0 && addressInt < 10
+            }
+            // [OP-041]
+            if (opcode.match(/^(EXT.*|CALL|CALLCODE|DELEGATECALL|STATICCALL)$/) != null) {
+                const idx = opcode.startsWith("EXT") ? 0 : 1
+                const addr = toAddress(log.stack.peek(idx).toString(16))
+                const addrHex = toHex(addr)
+                // this.debug.push('op=' + opcode + ' last=' + this.lastOp + ' stacksize=' + log.stack.length() + ' addr=' + addrHex)
+                if (this.currentLevel.contractSize[addrHex] == null && !isAllowedPrecompiled(addr)) {
+                    this.currentLevel.contractSize[addrHex] = {
+                        contractSize: db.getCode(addr).length,
+                        opcode
+                    }
+                }
+            }
+
+            // [OP-012]
```

Comment:
I love how we mention the specific rule that is being checked — great for reference! Could we do this in other places too?

Reply:
we have almost everywhere